python/private/python.bzl

```
@@ -267,11 +267,18 @@ def parse_modules(*, module_ctx, _fail = fail):
 def _python_impl(module_ctx):
     py = parse_modules(module_ctx = module_ctx)
 
-    # dict[str version, list[str] platforms]; where version is full
-    # python version string ("3.4.5"), and platforms are keys from
-    # the PLATFORMS global.
-    loaded_platforms = {}
-    for toolchain_info in py.toolchains:
+    # list of structs; see inline struct call within the loop below.
+    toolchain_impls = []
+
+    # list[str] of the base names of toolchain repos
+    base_toolchain_repo_names = []
+
+    # Create the underlying python_repository repos that contain the
+    # python runtimes and their toolchain implementation definitions.
+    for i, toolchain_info in enumerate(py.toolchains):
+        is_last = (i + 1) == len(py.toolchains)
```

```suggestion
    all_toolchains = zip(
        [False] * len(py.toolchains),
        py.toolchains,
    )[:-1]
    all_toolchains.append(
        (True, py.toolchains[-1])
    )
    for set_python_toolchain_constraint, toolchain_info in all_toolchains:
```

You could do this, I am curious what is more readable.

Also, I can't remember if the last toolchain is still special because we always know the `python_version` flag setting, so the comment/logic may be outdated. At least when I was bugfixing code in this location I think that the case when the `python_version` does not match the preceding entries never happens in bzlmod.

---

re: if no-version occurs: Good point. It would occur if...hmm...If --python_version was set to a version that wasn't known? This could occur if someone registered a toolchain outside of python.toolchain()